File: server/src/modules/task/overdueReminder.job.ts
import prisma from '@/core/prisma';
import { messageService } from '@/modules/message/message.service';

const REMINDER_EVENT_TYPE = 'OVERDUE_REMINDER';
const REMINDER_TITLE = '任务已逾期';

const ACTIVE_STATUSES = [
  'PENDING',
  'IN_PROGRESS',
  'SELF_TESTING',
  'QA_REVIEW',
  'REJECTED',
  'PAUSED'
] as const;

function formatDueDate(d: Date) {
  return d.toLocaleString('zh-CN', { hour12: false });
}

let started = false;

export function startOverdueReminderJob() {
  if (started) return;
  started = true;

  // 每 30 分钟跑一次；真正的“是否已提醒过”由 TaskTimeline 做幂等控制
  const intervalMs = 30 * 60 * 1000;
  void runOnce();

  const timer = setInterval(() => {
    void runOnce();
  }, intervalMs);

  // 允许 Node 在无其他事件时退出（例如测试/脚本场景）
  timer.unref?.();
}

async function runOnce() {
  const now = new Date();

  // 避免扫描太老的历史任务，降低 DB 压力
  const scanCutoff = new Date(now.getTime() - 14 * 24 * 60 * 60 * 1000);

  const todayStart = new Date(now);
  todayStart.setHours(0, 0, 0, 0);

  const tasks = await prisma.task.findMany({
    where: {
      dueDate: {
        not: null,
        lt: now,
        gte: scanCutoff
      },
      status: { in: ACTIVE_STATUSES as any }
    },
    select: {
      id: true,
      title: true,
      projectId: true,
      dueDate: true,
      mainAssigneeId: true,
      testerId: true,
      coAssignees: { select: { userId: true } },
      project: { select: { name: true } }
    }
  });

  if (!tasks.length) return;

  const taskIds = tasks.map((t) => t.id);

  const reminded = await prisma.taskTimeline.findMany({
    where: {
      taskId: { in: taskIds },
      eventType: REMINDER_EVENT_TYPE,
      createdAt: { gte: todayStart }
    },
    select: { taskId: true }
  });

  const remindedSet = new Set<number>(reminded.map((r) => r.taskId));

  // 批量顺序发送，避免瞬时推送风暴；失败不影响后续任务
  for (const task of tasks) {
    if (remindedSet.has(task.id)) continue;

    const due = task.dueDate ? new Date(task.dueDate) : null;
    const dueText = due ? formatDueDate(due) : '未知截止时间';
    const projectName = task.project?.name ?? '';

    // 1) 写入时间线：确保“每天每任务只提醒一次”
    await prisma.taskTimeline.create({
      data: {
        taskId: task.id,
        eventType: REMINDER_EVENT_TYPE,
        title: REMINDER_TITLE,
        content: `截止时间：${dueText}（项目：${projectName || '未命名'}）`,
        operatorId: null,
        fromStatus: null,
        toStatus: null,
        payload: { dueDate: due?.toISOString() ?? null }
      }
    });

    // 2) 站内消息：推给主负责人 + 验收人 + 协助人（排除操作者；Job 为系统消息，senderId=null）
    const receiverIds = new Set<number>();
    if (task.mainAssigneeId) receiverIds.add(task.mainAssigneeId);
    if (task.testerId) receiverIds.add(task.testerId);
    for (const ca of task.coAssignees ?? []) {
      if (ca.userId) receiverIds.add(ca.userId);
    }

    if (!receiverIds.size) continue;

    await Promise.allSettled(
      Array.from(receiverIds).map((receiverId) =>
        messageService.sendRealTimeMessage(
          {
            receiverId,
            messageType: 'TASK',
            title: REMINDER_TITLE,
            content: projectName
              ? `任务「${task.title}」已逾期，截止时间：${dueText}（项目：${projectName}）。请尽快处理。`
              : `任务「${task.title}」已逾期，截止时间：${dueText}。请尽快处理。`,
            extra: {
              biz: 'task',
              action: 'overdue',
              taskId: task.id,
              projectId: task.projectId,
              dueDate: due?.toISOString() ?? null
            }
          },
          null
        )
      )
    );
  }
}

File: java-server/src/main/java/com/todothink/modules/task/job/OverdueReminderJob.java
package com.todothink.modules.task.job;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.todothink.modules.message.dto.SendMessageRequest;
import com.todothink.modules.message.service.MessageService;
import com.todothink.modules.task.entity.Project;
import com.todothink.modules.task.entity.Task;
import com.todothink.modules.task.entity.TaskCoAssignee;
import com.todothink.modules.task.entity.TaskTimeline;
import com.todothink.modules.task.mapper.ProjectMapper;
import com.todothink.modules.task.mapper.TaskCoAssigneeMapper;
import com.todothink.modules.task.mapper.TaskMapper;
import com.todothink.modules.task.mapper.TaskTimelineMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 逾期任务提醒定时任务（对齐 Node overdueReminder.job.ts）。
 */
@Component
public class OverdueReminderJob {

    private static final Logger log = LoggerFactory.getLogger(OverdueReminderJob.class);

    private static final String REMINDER_EVENT_TYPE = "OVERDUE_REMINDER";
    private static final String REMINDER_TITLE = "任务已逾期";

    private static final List<String> ACTIVE_STATUSES = Arrays.asList(
            "PENDING", "IN_PROGRESS", "SELF_TESTING", "QA_REVIEW", "REJECTED", "PAUSED");

    private static final DateTimeFormatter DUE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private final TaskMapper taskMapper;
    private final TaskTimelineMapper taskTimelineMapper;
    private final TaskCoAssigneeMapper taskCoAssigneeMapper;
    private final ProjectMapper projectMapper;
    private final MessageService messageService;
    private final ObjectMapper objectMapper;

    public OverdueReminderJob(TaskMapper taskMapper, TaskTimelineMapper taskTimelineMapper,
            TaskCoAssigneeMapper taskCoAssigneeMapper, ProjectMapper projectMapper,
            MessageService messageService, ObjectMapper objectMapper) {
        this.taskMapper = taskMapper;
        this.taskTimelineMapper = taskTimelineMapper;
        this.taskCoAssigneeMapper = taskCoAssigneeMapper;
        this.projectMapper = projectMapper;
        this.messageService = messageService;
        this.objectMapper = objectMapper;
    }

    @Scheduled(fixedDelay = 30 * 60 * 1000L, initialDelay = 60 * 1000L)
    public void run() {
        try {
            runOnce();
        } catch (Exception e) {
            log.error("[overdue] reminder job failed: {}", e.getMessage(), e);
        }
    }

    private void runOnce() {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime scanCutoff = now.minusDays(14);
        LocalDateTime todayStart = now.toLocalDate().atStartOfDay();

        List<Task> tasks = taskMapper.selectList(new LambdaQueryWrapper<Task>()
                .isNotNull(Task::getDueDate)
                .lt(Task::getDueDate, now)
                .ge(Task::getDueDate, scanCutoff)
                .in(Task::getStatus, ACTIVE_STATUSES));

        if (tasks.isEmpty()) {
            return;
        }

        List<Integer> taskIds = tasks.stream().map(Task::getId).collect(Collectors.toList());

        List<TaskTimeline> reminded = taskTimelineMapper.selectList(new LambdaQueryWrapper<TaskTimeline>()
                .in(TaskTimeline::getTaskId, taskIds)
                .eq(TaskTimeline::getEventType, REMINDER_EVENT_TYPE)
                .ge(TaskTimeline::getCreateTime, todayStart)
                .select(TaskTimeline::getTaskId));

        Set<Integer> remindedSet = new HashSet<Integer>();
        for (TaskTimeline row : reminded) {
            remindedSet.add(row.getTaskId());
        }

        for (Task task : tasks) {
            if (remindedSet.contains(task.getId())) {
                continue;
            }

            LocalDateTime due = task.getDueDate();
            String dueText = due != null ? DUE_FORMAT.format(due) : "未知截止时间";
            String projectName = loadProjectName(task.getProjectId());

            TaskTimeline timeline = new TaskTimeline();
            timeline.setTaskId(task.getId());
            timeline.setEventType(REMINDER_EVENT_TYPE);
            timeline.setTitle(REMINDER_TITLE);
            timeline.setContent("截止时间：" + dueText + "（项目：" + (projectName.length() > 0 ? projectName : "未命名") + "）");
            timeline.setOperatorId(null);
            timeline.setFromStatus(null);
            timeline.setToStatus(null);
            try {
                Map<String, Object> payload = new HashMap<String, Object>();
                payload.put("dueDate", due);
                timeline.setPayload(objectMapper.writeValueAsString(payload));
            } catch (Exception e) {
                timeline.setPayload(null);
            }
            taskTimelineMapper.insert(timeline);

            Set<Integer> receiverIds = new HashSet<Integer>();
            if (task.getMainAssigneeId() != null) {
                receiverIds.add(task.getMainAssigneeId());
            }
            if (task.getTesterId() != null) {
                receiverIds.add(task.getTesterId());
            }
            List<TaskCoAssignee> coRows = taskCoAssigneeMapper.selectList(new LambdaQueryWrapper<TaskCoAssignee>()
                    .eq(TaskCoAssignee::getTaskId, task.getId())
                    .select(TaskCoAssignee::getUserId));
            for (TaskCoAssignee ca : coRows) {
                if (ca.getUserId() != null) {
                    receiverIds.add(ca.getUserId());
                }
            }
            if (receiverIds.isEmpty()) {
                continue;
            }

            String content = projectName.length() > 0
                    ? "任务「" + task.getTitle() + "」已逾期，截止时间：" + dueText + "（项目：" + projectName + "）。请尽快处理。"
                    : "任务「" + task.getTitle() + "」已逾期，截止时间：" + dueText + "。请尽快处理。";

            for (Integer receiverId : receiverIds) {
                try {
                    SendMessageRequest msg = new SendMessageRequest();
                    msg.setReceiverId(receiverId);
                    msg.setMessageType("TASK");
                    msg.setTitle(REMINDER_TITLE);
                    msg.setContent(content);
                    Map<String, Object> extra = new HashMap<String, Object>();
                    extra.put("biz", "task");
                    extra.put("action", "overdue");
                    extra.put("taskId", task.getId());
                    extra.put("projectId", task.getProjectId());
                    extra.put("dueDate", due);
                    msg.setExtra(extra);
                    messageService.sendRealTimeMessage(msg, null);
                } catch (Exception e) {
                    log.debug("[overdue] send message failed taskId={} receiverId={}", task.getId(), receiverId, e);
                }
            }
        }
    }

    private String loadProjectName(Integer projectId) {
        if (projectId == null) {
            return "";
        }
        Project project = projectMapper.selectOne(new LambdaQueryWrapper<Project>()
                .eq(Project::getId, projectId)
                .select(Project::getName));
        return project != null && project.getName() != null ? project.getName() : "";
    }
}
